### ISSUE TITLE
deserialise_snapshot() assumes that it runs single-threaded, but frontends are already open

### ISSUE THREAD

**Issue by achamayou** [microsoft/CCF#5141]
Transactions executed while the snapshot is being deserialised may see an inconsistent view of the KV (state not in sync with version, or version not in sync with history).

The frontends should either stay closed, or only emit transactionless commands until the snapshot is fully deserialised, to avoid a giant lock on node state.

**Comment by achamayou** [comment]
# Investigation summary: CCF issue #5141

**Issue:** [microsoft/CCF#5141 - `deserialise_snapshot()` assumes that it runs single-threaded, but frontends are already open](https://github.com/microsoft/CCF/issues/5141)  
**Investigated:** 2026-07-13  
**Current `main` reviewed:** [`74cdc0a6797981d463dd58d0d27ab663041aafab`](https://github.com/microsoft/CCF/commit/74cdc0a6797981d463dd58d0d27ab663041aafab)

## Public history

The issue was opened on 2023-03-28 over transactions potentially observing KV state, store version, and history at different stages of snapshot deserialisation. It remains open, with no comments, assignee, or closing pull request.

Two merged PRs are cross-referenced:

- [#5142](https://github.com/microsoft/CCF/pull/5142) serialised access to the KV map set and made transactions spanning snapshot installation fail and retry rather than freely using incompatible map state.
- [#5317](https://github.com/microsoft/CCF/pull/5317) removed a transactional KV read from the frontend-open check, reducing deadlock risk and ensuring modern member and user frontends are opened by post-snapshot hooks.

Neither PR declares that it fixes #5141.

## How the issue arose

### PR #5142 exposed the single-threading assumption

[#5142](https://github.com/microsoft/CCF/pull/5142) added `maps_lock` acquisition to `Store::get_map()` and introduced `get_map_unsafe()` for callers already holding the lock. It moved history initialisation outside the `maps_lock` scope because `TxHistory::init_from_snapshot()` reads a KV table and would otherwise recursively call the newly locking `get_map()`.

The review records the assumption directly:

- The author wrote that this was safe because `deserialise_snapshot()` was "[supposed to be] called in single threaded way" ([discussion](https://github.com/microsoft/CCF/pull/5142#discussion_r1150843568)).
- The issue author replied: "To be clear, it's not now, but that's captured in #5141" ([discussion](https://github.com/microsoft/CCF/pull/5142#discussion_r1151712661)).

The resulting lock boundary and post-lock history initialisation remain in current `main`.

### PR #5317 removed one dangerous frontend read

Before [#5317](https://github.com/microsoft/CCF/pull/5317), `RpcFrontend::is_open(kv::Tx&)` could query the service table while holding `open_lock`. The PR replaced that lazy transactional check with a simple `open()`/`is_open()` flag.

This stopped a closed user frontend from probing the KV merely to determine whether it was open. Current join code also invokes snapshot-produced global hooks only after `deserialise_snapshot()` returns, so member and user frontends open after history restoration. It did not change the early-open node frontend or make snapshot publication atomic.

## Current implementation

### Map state and version are published before history

Current [`Store::deserialise_snapshot()`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/kv/store.h#L393-L535):

1. takes `maps_lock` and locks existing maps;
2. deserialises and applies every map;
3. writes `version = v` and `last_replicated = v`;
4. releases `maps_lock`; and
5. calls `TxHistory::init_from_snapshot()`.

Current [`HashedTxHistory::init_from_snapshot()`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/node/history.h#L719-L752) first reads `SERIALISED_MERKLE_TREE` through a KV transaction, then takes `history.state_lock`, asserts that the replicated-state tree is empty, deserialises it, and appends the snapshot transaction hash.

No lock or readiness barrier spans map/version publication and completion of history initialisation. A history-initialisation failure is also non-transactional: `deserialise_snapshot()` returns `FAIL` after map state and store version have already changed.

### The node frontend is open during snapshot installation

[`NodeState::create()`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/node/node_state.h#L973-L995) accepts node TLS connections and opens `ActorsType::nodes` before `setup_history()`, `setup_snapshotter()`, and `setup_encryptor()`. Joining nodes apply the snapshot later during join-response processing ([`node_state.h`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/node/node_state.h#L1457-L1495)).

The node frontend includes both transactionless command endpoints and KV-backed endpoints. `RpcFrontend` creates a `CommittableTx` before endpoint dispatch ([`frontend.h`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/node/rpc/frontend.h#L650-L702)); command endpoints leave it unused, while endpoints such as `/node/state`, `/node/network`, and `/node/network/nodes` acquire KV handles.

### Worker threads make overlap possible

`worker_threads` defaults to zero but is configurable. The host starts `worker_threads + 1` enclave threads, and join-response handling schedules node-state processing, including large-snapshot deserialisation, as a task. With workers enabled, snapshot application can overlap incoming RPC work.

## Concrete interleavings

### A. A transaction captures the old version, then waits for new maps

1. An RPC transaction records the pre-snapshot version.
2. Snapshot deserialisation owns `maps_lock`, resets each map to snapshot state, and advances the store version.
3. The RPC blocks in `get_map()` until `maps_lock` is released.
4. It asks the post-snapshot map for a change set at the pre-snapshot version.

Current code generally fails safely here: `Map::create_change_set()` returns null when the requested version predates the snapshot roll, `BaseTx` raises `CompactedVersionConflict`, and `RpcFrontend` retries. This materially mitigates the original state/version interleaving for normal RPCs.

### B. A transaction sees the new version before history exists

1. Snapshot map changes are installed.
2. `Store::version` is set to the snapshot version.
3. `maps_lock` is released.
4. Before `init_from_snapshot()` completes, another thread reads the new KV state or appends a newly committed entry to history.

This leaves two structurally possible outcomes:

- a history or root reader observes a history whose end index does not match the published KV version;
- a concurrent `append_entry()` populates the empty tree before `init_from_snapshot()` checks it, triggering `CCF_ASSERT(!replicated_state_tree.in_range(1))`.

`history.state_lock` serialises the history operations themselves but does not impose the required ordering between KV publication and history initialisation.

## Fixed, mitigated, and unresolved areas

| Area | Status |
|---|---|
| Unsynchronised access to the map collection | **Fixed/mitigated by #5142.** `get_map()` takes `maps_lock`. |
| RPC spanning pre- and post-snapshot map rolls | **Mitigated.** It normally raises `CompactedVersionConflict` and retries. |
| User frontend probing the service table while closed | **Fixed by #5317.** Open state no longer requires a transactional read. |
| Member/user frontend opening during modern join snapshot application | **Largely mitigated.** Snapshot hooks run after deserialisation. |
| Node frontend open during join/recovery snapshot application | **Unresolved.** It opens before history setup and includes KV-backed endpoints. |
| Atomic publication of maps, version, and history | **Unresolved.** History is initialised after map/version publication. |
| Direct regression coverage for concurrent frontend traffic during deserialisation | **Missing.** |

## Test coverage and blind spots

Current tests cover the pieces only in sequence:

- [`kv_snapshot.cpp`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/kv/test/kv_snapshot.cpp#L255-L309) creates a transaction before synchronously applying a snapshot and checks that its later commit conflicts. It does not run the transaction and deserialiser concurrently or attach history.
- [`node/test/snapshot.cpp`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/src/node/test/snapshot.cpp#L95-L165) verifies sequential Merkle-history restoration.

Neither test uses a worker thread, frontend RPC, barrier inside history initialisation, or assertion about frontend readiness. Current [`tsan_env_suppressions`](https://github.com/microsoft/CCF/blob/74cdc0a6797981d463dd58d0d27ab663041aafab/tsan_env_suppressions#L7-L12) also suppresses deadlock reports in `store.h`/`untyped_map.h` and a race in `node/rpc/frontend.h`.

## Recommended resolution

1. **Add a startup-store readiness barrier.** Frontend dispatch should check it before creating or seeding a KV transaction. Publish readiness only after snapshot maps, store version, history, consensus, and snapshot hooks are coherent.
2. **Keep the node frontend command-only until readiness.** TLS can remain available for bootstrap, but KV-backed node endpoints should return a deterministic `503`/`FrontendNotOpen`.
3. **Avoid extending one giant lock across history and frontend setup.** `init_from_snapshot()` reads KV, and history/KV lock inversion is already a concern. If concurrent availability is required, build a complete Store+History off to the side and atomically publish it, or use a generation barrier that restarts pre-publication transactions.
4. **Add a deterministic multithreaded regression test.** Pause immediately before history initialisation, issue transactionless and KV-backed node requests with `worker_threads >= 1`, and verify that only transactionless endpoints succeed before readiness and that KV version and history agree after release.
5. **Run the test under TSAN and narrow the suppressions.**
6. **Cover every snapshot path:** normal join, public-only recovery, full recovery-store initialisation, local snapshots, and runtime-fetched snapshots.
